compiler/plugins/target/AMD-AIE/aie/AMDAIEObjectFifoStatefulTransform.cpp

```diff
@@ -284,589 +201,460 @@ void createDMA(
-    if (static_cast<size_t>(numRel) > acquiredIndices.size())
-      llvm::report_fatal_error(
-          "cannot release more elements than are "
-          "already acquired");
-    for (int i = 0; i < numRel; i++)
-      acquiredIndices.erase(acquiredIndices.begin());
+  const ObjectFifoEndpointResource &endpointResource =
+      port == ObjectFifoPort::Produce
+          ? resourceMap.at(op).producerResource
+          : resourceMap.at(op).consumerResources.at(tileOp);
+
+  builder.setInsertionPointAfter(acquireOp);
+  if (failed(createUseLocks(builder, op, port, acquireOp.getSize(),
+                            LockAction::AcquireGreaterEqual,
+                            endpointResource))) {
+    return failure();
   }
 
-  // acquire locks
-  size_t numLocks = acquireOp.getSize();
-  size_t alreadyAcq = acquiredIndices.size();
-  size_t numCreate = numLocks > alreadyAcq ? numLocks - alreadyAcq : 0;
-
-  {
-    OpBuilder::InsertionGuard gg(builder);
-    builder.setInsertionPointAfter(acquireOp);
-    createUseLocks(builder, op, port, acqPerFifo, numCreate,
-                   LockAction::AcquireGreaterEqual, objFifoLinks, locksPerFifo);
+  for (Operation *userOp : acquireOp->getUsers()) {
+    auto subviewAccessOp = dyn_cast<ObjectFifoSubviewAccessOp>(userOp);
+    if (!subviewAccessOp) {
+      return acquireOp.emitOpError()
+             << "currently only supports `aie.objectfifo.subview.access` users";
+    }
+    size_t index = subviewAccessOp.getIndex();
+    size_t bufferIndex =
+        (createOpToIndex[op] + index) % endpointResource.buffers.size();
+    BufferOp bufferOp = endpointResource.buffers[bufferIndex];
+    subviewAccessOp.getResult().replaceAllUsesWith(bufferOp.getResult());
   }
+  // Increment index to rotate through available buffers objectFifo acquires.
+  createOpToIndex[op] += acquireOp.getSize();
+  return success();
+}
 
-  // if objFifo was linked with others, find which objFifos
-  // elements to use
-  ObjectFifoCreateOp target = op;
-  if (linkOp && objFifoLinks.contains(*linkOp))
-    target = objFifoLinks.at(*linkOp);
-
-  // create subview: buffers that were already acquired + new acquires
-  for (int i = 0; i < numCreate; i++) {
-    acquiredIndices.push_back(start);
-    start = (start + 1) % objFifoSize(op);
+/// Utility to create a vector of buffer ops for an objectFifo.
+SmallVector<BufferOp> createBuffers(OpBuilder &builder,
+                                    const AMDAIEDeviceModel &deviceModel,
+                                    ObjectFifoCreateOp createOp,
+                                    size_t numBuffers, TileOp tile,
+                                    const std::string &prefix, size_t index) {
+  SmallVector<BufferOp> buffers;
+  if (deviceModel.isShimTile(tile.getCol(), tile.getRow())) return buffers;
+  auto fifoType = llvm::cast<AIEObjectFifoType>(createOp.getElemType());
+  auto elemType = llvm::cast<MemRefType>(fifoType.getElementType());
```

Comment:
Can we get rid of `llvm::` here and elsewhere? I believe `cast<...>` is included in the default mlir namespace ?

Reply:
Done